### PR TITLE
Use Relational Data Modeling Techniques to design the content of the Work Zone Data Specification

### PR DESCRIPTION
**Pull request by DeraldDudley**
**What:**  Use Relational Data Modeling Techniques (RDMT) to model Work Zone Data Specification content.

**Adopting Pull Request #50, also adopts following pull request...**
- #44 Mobility Impact and Lane Level Impacts + Restrictions
- #45 Lane Enumerations
- #48 Adding Mobility Impact Only

**Proposed Changes**
- All Proposed Changes can be found in the [data tables directory](https://github.com/DeraldDudley/jpo-wzdx/tree/master/data-tables) of Derald Dudley's forked [copy of the jpo-wzdx repository](https://github.com/DeraldDudley/jpo-wzdx)

- [Logical Data Model](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_erd.png) (Entity Relationship Data Model)
- The [WZDx Header](https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/data-tables/wzdx-header-data.md) will be converted to [Road Event Feed Info Table](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_feed_info.md)
- The [Common Core Data Dictionary](https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/data-tables/common-core-dictionary.md) will be converted to the [Road Events Table](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_events.md)

**New Tables to Accommodate Lane Types, Closures, Effect on Traveled Path, Lane Restrictions**
- [Activities](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/activities.md)
- [Lanes, Closures, and Effect of Traveled Path](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/lanes.md)
- [Lane Restrictions](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/lane_restrictions.md)

**Why:**  Version 1 of the Work Zone Data Specification relies on data dictionary, markup, and exchange languages (TMDD, XML, JSON) to define content and formatting.  It does not employ well established data design methods to define objects, object properties, or relationships.

Using Relational Data Modeling Techniques to design the Work Zone Data Specification will;
- Organize the data more efficiently
- Increase consistency
- Ease data retrieval
- Enable a flexible and scalable data model
- Promote Standards-Centric development of applications

Using Relational Data Modeling Techniques to design the Work Zone Data Specification will not;
- Dictate how data is managed at the source
- Change the specifications content
- Prohibit publishing content according to the TMDD
- Prohibit publishing content according to version 1 of the specification

### CODE REVIEW

**Comment by DeraldDudley** [comment]
20191021 @ 1425 - Updated PR with a Normalized Common Core Table.

**Comment by k-vaughn** [comment]
It should be noted that ISO TC 204 (Intelligent Transport Systems) WG 1 (Architecture) is leading an effort to create an ITS-wide logical data model for the explicit purpose to define how data from different protocols can be shared across protocols, domains, etc. We are currently working with the Smart City and Internet of Things community to ensure our approach to defining this data model is consistent with those much larger efforts that we will eventually need to link to. 

There are also ongoing discussions about how these data models might be linked into the National ITS Architecture (http://arc-it.org).

It would be useful if any data models produced by the WZDx team were coordinated with these broader efforts. This likely suggests using UML as the data modeling language and eventually producing traces from each element within the WZDx model (which would be a "physical data model") to the ITS logical data model. These efforts are just getting started and are currently hosted at https://github.com/k-vaughn/its-reference-model (likely to be moved to the ISO-TC204  page in the future)

**Comment by DeraldDudley** [comment]
@k-vaughn - Happy to model in UML.  Hope to have it done my Wednesday.

**Comment by CraigMoore-Sea** [comment]
The rich text formatting for road_events (Formerly Common Core Data Dictionary) table is not working. We need to get this clean up before voting. It's very hard to follow in the current state.

**Comment by DeraldDudley** [comment]
Yes.  I’m trying to clean it up this afternoon.

Derald Dudley
NGDA Transportation<https://communities.geoplatform.gov/ngda-transportation/> Theme Lead
FGDC Transportation Subcommittee<https://communities.geoplatform.gov/ngda-transportation/transportation-subcommittee/> Chair
Phone: 202.309.0940
Pronouns: he/him/his

The Office of Spatial Analysis and Visualization<http://maps.bts.dot.gov/>
Bureau of Transportation Statistics<http://www.rita.dot.gov/bts/>
Office of the Assistant Secretary of Transportation for Research and Technology<http://www.rita.dot.gov/>
United States Department of Transportation<http://www.transportation.gov/>

**Comment by CraigMoore-Sea** [comment]
I suggest we define the allowed geometry types to be LineString, MultiPoint, MultiLineString, and GeometryCollection; any collection must at least include a MultiPoint or Linestring feature.

**Comment by CraigMoore-Sea** [comment]
Consider adding new travel_path_effect field to the lanes table. We would move the shift-left and shift-right values to this table from the lane_status field. We would add the following values:
•	alternating-one-way
•	merge
•	merge-left
•	merge-right	
•	shift
•	unknown

We may want to add an open-with-restrictions value to the lane status field. This would serve as notice that related restrictions records exist.

**Comment by DeraldDudley** [comment]
> Consider adding new travel_path_effect field to the lanes table. We would move the shift-left and shift-right values to this table from the lane_status field. We would add the following values:
> • alternating-one-way
> • merge
> • merge-left
> • merge-right
> • shift
> • unknown
> 
> We may want to add an open-with-restrictions value to the lane status field. This would serve as notice that related restrictions records exist.

Would shift, shift-left, and shift-right be more explicit?

**Comment by j-d-b** [comment]
I'm in the process of reviewing this currently.

Overall, I fear that not many people will have time to review in the depth that they should considering several new commits are being pushed every day. I support the relational data model conceptually and believe it should be implemented in the v2 release as now is the time for large changes like this. However, I think this PR is making too many changes (new commits every hour today, one day before voting [Nov 18]) and is thus not yet stable enough to review and be ready for voting by tomorrow (Nov 19). If I had to vote now, I would vote "no" for this reason, not because I don't agree with the idea or your implementation.

Additionally, I don't think it was a good decision to address other issues (#44, #45, #48) in this PR as it broadens the PR's scope and doesn't allow for voters to accept some of these and not others. This PR should be specific to the restructuring/relational data model until it is approved, after which it can implement the other issues.

Relatedly, there was much discussion followed by a passing vote on #33 to define the enumerated types (called `activity_type` in this relational data PR) for a work zone `event_type`, however, that doesn't seem to be referenced here. Even `event_type`, which is critical to this PR being completed, is not yet defined. Ideally, defining that would require some discussion which doesn't seem possible given the current timeline (one day until voting). There may be more "missing pieces" like this; when I finish my review I will note all that I have found here.

**Comment by j-d-b** [comment]
I've completed my review and compiled my comments below:

## Actions
* The `activity_type` enumerations for a "Work Zone" `event_type` should be included in (`enumerated_field_definitions.md`)[https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/enumerated_field_definitions.md] as approved in #33.
*The full `event_type` enumeration needs to be defined, though potentially non-blocking for WZDx v2. Relatedly, an `activity_type` enumeration for each `event_type` will be necessary. These should be a future item for completeness and adoption of this structure in other contexts.
* The `full-spec` and child `full-spec.md` should be removed, as they are no longer relevant and all `data-tables` references no longer exist.
* The content of the `create-feed` directory needs to be updated. Having examples (e.g. JSON and XML) of the new models/field names in use is important; this information was critical for me when first getting to understand the WZDx v1.1 spec. 
* The content of the `sample-files` directory needs to be updated.
* The project root `README.md` references to `full-spec.md` and “common core” should be edited/removed, as both are no longer relevant.
* Fix typo in the [ERD diagram](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_erd.png); `event_subidentifieer` to `event_subidentifier`.

## Comments
* The `road_name` and `road_number` fields are “new” in the sense that before they were part of the [BeginLocation data frame in v1.1](https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/data-tables/data-frames.md#beginlocation). This approach has a flaw in that the user can't specify if the EndLocation roadway is different than the StartLocation. However, it was at least clear at which point in the road event the road name referred to. Following the changes proposed in this PR, if the road event spans multiple roadways, it will be unclear where the `road_name` and `road_number` refer to. There are a few ways to solve this (see #52), but ideally, the solution would be the result of its own discussion and this PR would just match v1.1. To do that, these fields would be chang to `start_road_name` and `start_road_number`.
* The description of `event_begin_mp_demarcation` and `event_end_mp_demarcation` in the [road events](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_events.md) table is inconsistent with the field name.
* There is inconsistency between the ERD diagram `road_event_feed_info` in the diagram has `event_issuing_organization` while the [data table](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_feed_info.md) uses `feed_publisher`.
* In v1.1, `issuingOrganzation` was a field on the road event and in this PR it has been moved to the `road_event_feed_info` table which was previously the header. This means that road events cannot have different issuers which may be fine (and is cleaner, if it’s fine) but is a larger change than just being part of the relational data model. Ideally, all breaking changes like this which were included as part of this PR should be separate PRs and up to their own discussion and voting.
* It would be very helpful if the enumerations for each enumerated type were hyperlinked to from their respective data table.
* Overall, this is a lot cleaner—more consistent naming and field type clarity—than v1.1.

## Questions
* What is the source of the “demarcation” terminology and fields in v1.1?
* Why is v1.1’s `roadDirection` renamed to `event_direction` when `roadName` and `roadNum` translate pretty directly to `road_number` and `road_name`?
* I think it would be helpful to have a “migration guide” which indicates the mapping between new fields in this PR to the old common core data.

Let me know if you need any clarification on any of this. I see

**Comment by natedeshmukhtowery** [comment]
Hi Jacob - thanks for this super comprehensive review. We agree with your suggestions on a streamlined approach for review and will put that forward shortly. Thank you for weighing in, and look forward to continued discussion.

**Comment by DeraldDudley** [comment]
> I've completed my review and compiled my comments below:
> 
> ## Needs Action
> 1. The `activity_type` enumerations for a "Work Zone" `event_type` should be included in [`enumerated_field_definitions.md`](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/enumerated_field_definitions.md) as approved in #33.
> 2. The full `event_type` enumeration needs to be defined, though potentially non-blocking for WZDx v2. Relatedly, an `activity_type` enumeration for each `event_type` will be necessary. These should be a future item for completeness and adoption of this structure in other contexts.
>3. The `full-spec` and child `full-spec.md` should be removed, as they are no longer relevant and all `data-tables` references no longer exist.
>4. The content of the `create-feed` directory needs to be updated. Having examples (e.g. JSON and XML) of the new models/field names in use is important; this information was critical for me when first getting to understand the WZDx v1.1 spec.
>5. The content of the `sample-files` directory needs to be updated.
>6. The project root `README.md` references to `full-spec.md` and “common core” should be edited/removed, as both are no longer relevant.
>7. Fix typo in the [ERD diagram](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_erd.png); `event_subidentifieer` to `event_subidentifier`.
> 
> ## Comments
>8. The `road_name` and `road_number` fields are “new” in the sense that before (v1.1) they were part of the [BeginLocation data frame](https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/data-tables/data-frames.md#beginlocation). This approach has a flaw in that the user can't specify if the EndLocation roadway is different than the StartLocation. However, it was at least clear at which point in the road event the road name referred to. Following the changes proposed in this PR, if the road event spans multiple roadways, it will be unclear where the `road_name` and `road_number` refer to. There are a few ways to solve this (see #52), but ideally, the solution would be the result of its own discussion and this PR would just match v1.1. To do that, these fields would be changed to `start_road_name` and `start_road_number`.
>9. The description of `event_begin_mp_demarcation` and `event_end_mp_demarcation` in the [road events](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_events.md) table is inconsistent with the field name.
>10. There is inconsistency between the ERD diagram `road_event_feed_info` in the diagram has `event_issuing_organization` while the [data table](https://github.com/DeraldDudley/jpo-wzdx/blob/master/data-tables/road_event_feed_info.md) uses `feed_publisher`.
>11.  In v1.1, `issuingOrganzation` was a field on the road event and in this PR it has been moved to the `road_event_feed_info` table which was previously the header. This means that road events cannot have different issuers which may be fine (and is cleaner, if it’s fine) but is a larger change than just being part of the relational data model. Ideally, all breaking changes like this which were included as part of this PR should be separate PRs and up to their own discussion and voting.
>12. It would be very helpful if the enumerations for each enumerated type were hyperlinked to from their respective data table.
>13. Overall, this is a lot cleaner—more consistent naming and field type clarity—than v1.1.
> 
> ## Questions
>14. What is the source of the “demarcation” terminology and fields in v1.1?
>15. Why is v1.1’s `roadDirection` renamed to `event_direction` when `roadName` and `roadNum` translate pretty directly to `road_number` and `road_name`?
>16. I think it would be helpful to have a “migration guide” which indicates the mapping between new fields in this PR to the old common core data.
> 
> Let me know if you need clarification on any of this.

Thank you @j-d-b !  

I've numbered the bullets for clarity. 

Actions
1. Activity Type Enumerations - Agree.
2. Work Type Enumerations - Agree.
- The only event_type being considered now is Work Zone; It is an enumeration of 1 value right but has room to grow.
- Agree that conditional Activity Types, dependent on Event Types, are necessary.
- **Are you proposing that activity types not be considered for this version?**
3. Removal of full_spec and full-spec.md - Agree
4. Update Create Feed Directory - Agree
5. Update sample-files directory - Agree
6. Update root README.md - Agree
7. Fix typo in ER Diagram - Good Catch, will do. Thank you.

Comments
8. Providing start_road_name, end_road_name, start_road_number, and end_road_number:
- One proposed solution to accommodate [Road Concurrency](https://en.wikipedia.org/wiki/Concurrency_(road)) included allowing a string of separated values to indicate multiple street names and numbers.
- E.g. "US12N; US70E; US220N; US421S; I40E; NC85N" or "Tamiami Trail; North Tamiami Trail; Cleveland Avenue"
- The proposed solution of a start_road_name, end_road_name, start_road_number, and end_road_number does not accommodate Road Concurrency and does not capture all road names along an event.
- Technically, to achieve normalization, the road names should be in a related table; this is probably a good place not to normalize the data.

9. The field names event_begin_mp_demarcation and event_end_mp_demarcation are awful. **I agree!**<br>
I changed the accuracy descriptor because the information being conveyed is not a measure of accuracy; Its an indication of whether or not the provided information have been verified. How do you feel about using a boolean is_verified flag for "accuracy" fields?
- event_beginning_is_verified
- event_ending_is_verified
- event_start_date_is_verified
- event_end_date_is_verified
- event_beginning_milepost_is_verified
- event_ending_milepost_is_verified

10. event_issuing_organization/feed_publisher discrepancy
- Good catch, thank you

11. In v1.1 issuingOrganzation was a field on the road event and in this PR it has been moved to the road_event_feed_info table which was previously the header. This means that road events cannot have different issuers.
- This is not true.  The road_event_info_feed table can accommodate more than one issuing organization.  The subsequent road events would be related to the road_event_info_feed by the feed_id.
- A JSON Example:
{
	"road_event_feed_info": [{
			"feed_id": 1,
			"issuing_organization": "State DOT",
			"timeStampUpdate": "time",
			"metadataURL": "url",
			"versionNo": "1"
		},
		{
			"feed_id": 2,
			"issuing_organization": "County DOT",
			"timeStampUpdate": "time",
			"metadataURL": "url",
			"versionNo": "1"
		},
		{
			"feed_id": 3,
			"issuing_organization": "Local DOT",
			"timeStampUpdate": "time",
			"metadataURL": "url",
			"versionNo": "1"
		}
	]
}
- The issuing_organization field belongs in the road_event_feed_info table because its a more efficient method of storing the data.

12. Hyperlinked enumeration to and from their respective data tables. - Agree

13. Overall, this is a lot cleaner—more consistent naming and field type clarity—than v1.1. - I agree

Questions
14. What is the source of the “demarcation” terminology and fields in v1.1?
- latitude-est, latitude-ver, longitude-est, and longitude-ver
- I was trying to find clarifying language to indicate if how location where marked.  
- Item 9 above is a better solution.
 
15. Why is v1.1’s roadDirection renamed to event_direction when roadName and roadNum translate pretty directly to road_number and road_name?
- Events like, work zones, are mapped in each direction.  So each event would have a single direction.
- Road name and number apply to the whole road and not just the event.

16. I think it would be helpful to have a “migration guide” which indicates the mapping between new fields in this PR to the old common core data. - I agree.
